File: torch_fizz.cc
/**
 * PyTorch FizzBuzz (Native code)
 *
 * This file implements FizzBuzz with native PyTorch Tensor code.
 * There are 2 implementations: one using regular torch::Tensor and the other using at::Tensor
 * These are referred to as PyTorch Native and PyTorch ATNative implementations.
 *
 * Please see the README.md for details.
 */
#include <iostream>
#include <chrono>
#include <torch/torch.h>
#include <torch/csrc/jit/import.h>

using namespace std;

struct FizzBuzz : public torch::nn::Module {
    FizzBuzz() {
    }

    torch::Tensor forward(torch::Tensor n) {
        int nmax = n.item().toInt();
        torch::Tensor fizz = torch::zeros(1);
        torch::Tensor buzz = torch::zeros(1);
        torch::Tensor fizzbuzz = torch::zeros(1);
        for (int i = 0; i < nmax; ++i) {
            if (i % 6 == 0) {
                fizzbuzz += 1;
            } else if (i % 3 == 0) {
                buzz += 1;
            } else if (i % 2 == 0) {
                fizz += 1;
            }
        }
        return torch::stack({fizz, buzz, fizzbuzz});
    }
};

struct FizzBuzzAT : public torch::nn::Module {
    FizzBuzzAT() {
    }

    torch::Tensor forward(at::Tensor n) {
        int nmax = n.item().toInt();
        at::Tensor fizz = at::zeros(1);
        at::Tensor buzz = at::zeros(1);
        at::Tensor fizzbuzz = at::zeros(1);
        for (int i = 0; i < nmax; ++i) {
            if (i % 6 == 0) {
                fizzbuzz += 1;
            } else if (i % 3 == 0) {
                buzz += 1;
            } else if (i % 2 == 0) {
                fizz += 1;
            }
        }
        return torch::stack({fizz, buzz, fizzbuzz});
    }
};

// C++ template to print vector container elements
template <typename T>
ostream& operator<<(ostream& os, const vector<T>& v)
{
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        os << v[i];
        if (i != v.size() - 1)
            os << ", ";
    }
    os << "]\n";
    return os;
}

vector<int> fizzbuzz(int n) {
    int fizz = 0, buzz = 0, fizzbuzz = 0;
    for (int i = 0; i < n; ++i) {
        if (i % 6 == 0) ++fizzbuzz;
        else if (i % 3 == 0) ++buzz;
        else if (i % 2 == 0) ++fizz;
    }
    return {fizz, buzz, fizzbuzz};
}

vector<int> fizzbuzz_volatile(volatile int n) {
    volatile int fizz = 0, buzz = 0, fizzbuzz = 0;
    for (int i = 0; i < n; ++i) {
        if (i % 6 == 0) ++fizzbuzz;
        else if (i % 3 == 0) ++buzz;
        else if (i % 2 == 0) ++fizz;
    }
    return {fizz, buzz, fizzbuzz};
}

int main(int argc, char* argv[]) {
    torch::NoGradGuard no_grad_guard;
    cout << "Loading model.\n";
    auto start = chrono::high_resolution_clock::now();
    torch::jit::script::Module model = torch::jit::load("/tmp/fizzbuzz.pyt");
    model.eval();
    auto end = chrono::high_resolution_clock::now();
    cout << "Time taken (PyTorch Load) (ms): "
        << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';

    start = chrono::high_resolution_clock::now();
    FizzBuzz cpp_model;
    cpp_model.eval();
    end = chrono::high_resolution_clock::now();
    cout << "Time taken (PyTorch Native Model Build) (ms): "
        << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';

    start = chrono::high_resolution_clock::now();
    FizzBuzzAT at_cpp_model;
    at_cpp_model.eval();
    end = chrono::high_resolution_clock::now();
    cout << "Time taken (PyTorch Native Model Build) (ms): "
        << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';


    for (int i = 0; i < 3; ++i) {
        cout << "\n\nRun: " << i << "\n";
        cout << "=========================================\n";
        auto n = torch::tensor(100000);
        start = chrono::high_resolution_clock::now();
        auto result = model.forward({n});
        end = chrono::high_resolution_clock::now();
        cout << "Result: [" << result << "]\n";
        cout << "Time taken (PyTorch Loaded) (ms): "
            << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';

        start = chrono::high_resolution_clock::now();
        result = cpp_model.forward({n});
        end = chrono::high_resolution_clock::now();
        cout << "Result: [" << result << "]\n";
        cout << "Time taken (PyTorch Native) (ms): "
            << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';

        start = chrono::high_resolution_clock::now();
        result = at_cpp_model.forward({n});
        end = chrono::high_resolution_clock::now();
        cout << "Result: [" << result << "]\n";
        cout << "Time taken (PyTorch ATNative) (ms): "
            << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';

        start = chrono::high_resolution_clock::now();
        auto vresult = fizzbuzz(100000);
        end = chrono::high_resolution_clock::now();
        cout << "Result: " << vresult;
        cout << "Time taken (C++ Native Run) (ms): "
            << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';
        cout << "Time taken (C++ Native Run) (usec): "
            << chrono::duration_cast<std::chrono::microseconds>(end - start).count() << '\n';

        start = chrono::high_resolution_clock::now();
        vresult = fizzbuzz_volatile(100000);
        end = chrono::high_resolution_clock::now();
        cout << "Result: " << vresult;
        cout << "Time taken (C++ Native Volatile) (ms): "
            << chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << '\n';
        cout << "Time taken (C++ Native Volatile) (usec): "
            << chrono::duration_cast<std::chrono::microseconds>(end - start).count() << '\n';
    }
    return 0;
}

File: torch_fizz.py
import time

import torch

class TorchFizzBuzz(torch.nn.Module):
    def __init__(self):
        super(TorchFizzBuzz, self).__init__()

    def forward(self, n: torch.Tensor):
        i = torch.tensor(0, dtype=torch.int32, requires_grad=False)
        fizz = torch.zeros(1)
        buzz = torch.zeros(1)
        fizzbuzz = torch.zeros(1)
        while i < n:
            if i % 6 == 0:
                fizzbuzz += 1
            elif i % 3 == 0:
                buzz += 1
            elif i % 2 == 0:
                fizz += 1
            i += 1
        return torch.stack([fizz, buzz, fizzbuzz])


class PyFizzBuzz:
    def model(self, n):
        fizz = 0
        buzz = 0
        fizzbuzz = 0
        for i in range(n):
            if i % 6 == 0:
                fizzbuzz += 1
            elif i % 3 == 0:
                buzz += 1
            elif i % 2 == 0:
                fizz += 1
        return [fizz, buzz, fizzbuzz]


torch.no_grad()
COUNT = 100000
n = torch.tensor(COUNT, dtype=torch.int32)

print("Saving PyTorch model.")
mod = TorchFizzBuzz()
# Convert to code
jit_script = torch.jit.script(mod)
print(jit_script)
print(jit_script.code)
torch.jit.save(jit_script, '/tmp/fizzbuzz.pyt')

mod = TorchFizzBuzz()
mod.eval()
start_ns = time.perf_counter_ns()
result = mod.forward(n)
end_ns = time.perf_counter_ns()
print("Result: ", result)
print("Time (PyTorch) (ms): ", (end_ns - start_ns)/1e6)

mod = TorchFizzBuzz()
mod.eval()
with torch.jit.optimized_execution(True):
    start_ns = time.perf_counter_ns()
    result = mod.forward(n)
    end_ns = time.perf_counter_ns()

print("Result: ", result)
print("Time (PyTorch optimized=True) (ms): ", (end_ns - start_ns)/1e6)


mod = TorchFizzBuzz()
mod.eval()
with torch.jit.optimized_execution(False):
    start_ns = time.perf_counter_ns()
    result = mod.forward(n)
    end_ns = time.perf_counter_ns()
print("Result: ", result)
print("Time (PyTorch optimized=False) (ms): ", (end_ns - start_ns)/1e6)

print("Loading PyTorch model.")
loaded_module = torch.jit.load('/tmp/fizzbuzz.pyt')
loaded_module.eval()
start_ns = time.perf_counter_ns()
result = loaded_module.forward(n)
end_ns = time.perf_counter_ns()
print("Result: ", result)
print("Time (PyTorch from Loaded) (ms): ", (end_ns - start_ns)/1e6)

pymod = PyFizzBuzz()
perf_counter_ns_start = time.perf_counter_ns()
result = pymod.model(COUNT)
perf_counter_ns_end = time.perf_counter_ns()
time_taken_ns = perf_counter_ns_end - perf_counter_ns_start
print('Result: ', result)
print('Time taken (Python3) (ms): ', time_taken_ns / 1e6)
